### DISCUSSION TITLE
Amasty Faq Integration

### DISCUSSION THREAD

**Discussion by IbrahimS2** [breezefront/community#264] [1. Help]
I am trying to get this module fully integrated, but I am hitting some walls, and I am not certain where I am failing:

```
<item name="Amasty_Faq/js/faq-collapsible" xsi:type="array">
    <!-- Path to our JS file -->
    <item name="path" xsi:type="string">Vendor_vendor/js/faq-collapsible</item>
    <!-- Config paths to check before injecting JS file -->
    <item name="enabled" xsi:type="helper" helper="Swissup\Breeze\Helper\Config::isEnabled">
        <param name="path">amastyfaq/general/enabled</param>
    </item>
    <!-- JS files to add before our file -->
    <item name="import" xsi:type="array">
        <item name="collapsible" xsi:type="string">Swissup_Breeze/js/components/ui/collapsible</item>
    </item>
</item>
```

```
define([
    'jquery',
    'collapsible'
], function ($) {

    $.widget('mage.amFaqCollapsible', $.mage.collapsible, {
        // Component alias to search in DOM structure
        component: 'amFaqCollapsible',

        _create: function () {
            // WCAG compatibility
            this.options.content = $(this.element).find(this.options.content);
            this._super();
        }
    });

    return $.mage.amFaqCollapsible;
});

```

It is not extending/working for some reason, any clue?

**Comment by vovayatsyuk** [comment]
Use breakpoints to see if everything works like it should.

$.mage.collapsible - is undefined in your case.

Try this approach:

```js
define([
    'jquery',
    'collapsible'
], function ($, collapsible) {
    'use strict';

    $.widget('mage.amFaqCollapsible', collapsible, {
        // Component alias to search in DOM structure
        component: 'amFaqCollapsible',

        _create: function () {
            // WCAG compatibility
            this.options.content = $(this.element).find(this.options.content);
            this._super();
        }
    });

    return $.mage.amFaqCollapsible;
});

```

**Reply by IbrahimS2** [reply]
@vovayatsyuk 
The first and the third errors are related to this integration, however, the second is related to elastic-suite and that is not working either.
![image](https://github.com/breezefront/community/assets/19901858/3b93a439-b7be-4f8b-aa32-99ee81f109db)

**Reply by IbrahimS2** [reply]
The module is dependent on https://github.com/magento/magento2/blob/2.4/lib/web/mage/collapsible.js

**Reply by vovayatsyuk** [reply]
These errors are caused by old breeze module. Update to the latest version and redeploy static content.

**Reply by IbrahimS2** [reply]
What breeze module specifically? I already have it up-to-date

**Reply by vovayatsyuk** [reply]
`$.breezemap.__register` function was added in swissup/module-breeze 2.14: https://github.com/breezefront/module-breeze/commit/61acc769752283eabc327edd8db5c1f17ffacc16

**Reply by IbrahimS2** [reply]
@IbrahimS2 You are right, but should not this dependency module updated upon updating the blank template? I just checked and it appears the module-breeze module is stuck at 2.13.1

From my composer:
![image](https://github.com/breezefront/community/assets/19901858/4e3a985a-f46b-465b-b769-169eeecfcb3e)

My stack:
PHP 8.2
Magento 2.4.6

**Reply by IbrahimS2** [reply]
@vovayatsyuk Please check if the Swissup composer is up-to-date because even the blank template is outdated. this is from the [Swissup composer JSON](https://ci.swissuplabs.com/api/packages.json) file:
![image](https://github.com/breezefront/community/assets/19901858/392e523e-67e1-4048-b67b-5efaeebb47cc)

**Reply by vovayatsyuk** [reply]
I will remove all breeze packages from the swissup repo because they are available from Packagist.

Try running `composer update "swissup/*" -W`

**Reply by IbrahimS2** [reply]
@vovayatsyuk It worked, thanks; I have done lots of integration with Breeze i.e. Mirasvit, Amasty and more; if I upload those to our public repo would you link to it on the integration page?

**Reply by vovayatsyuk** [reply]
Awesome! Sure I will!

**Reply by IbrahimS2** [reply]
@vovayatsyuk Can we have this widget instead of the one that comes with Swissup?
https://github.com/magento/magento2/blob/2.4/lib/web/mage/collapsible.js

**Reply by IbrahimS2** [reply]
@vovayatsyuk Same module, other issues:
![image](https://github.com/breezefront/community/assets/19901858/bb5d10c9-bbd9-4744-bf54-f0ad76b19046)
![image](https://github.com/breezefront/community/assets/19901858/05cf92e9-ed5f-4e76-811a-d1079a75701f)
![image](https://github.com/breezefront/community/assets/19901858/58afad0e-ace7-428b-9782-ad9aed03aa42)

Any clue?

**Reply by vovayatsyuk** [reply]
mageUtils is not available in breeze. You can copy its implementation in your module. Example:

```xml
<item name="Amasty_Faq/js/faq-collapsible" xsi:type="array">
    ...
    <item name="import" xsi:type="array">
        <item name="mageUtils" xsi:type="string">Vendor_Module/js/mage-utils</item>
    </item>
</item>
```

**mage-utils.js**:

```js
if (!$.breezemap.mageUtils)
    $.breezemap.mageUtils = {
        template: ...,
        extend: ...
    };
}
```

**Reply by IbrahimS2** [reply]
@vovayatsyuk I realized you do have this library: https://github.com/breezefront/module-breeze/blob/f6d12ad9c77a33db9e07a69a42c0e5003f0a5a74/view/frontend/web/js/core/template.js

Shouldn't it be enough? Is there any documentation of what you suggested in detail so I can know how to use breezemap more efficiently?

**Reply by vovayatsyuk** [reply]
We don't have a docs for this.

Breezemap is a registry of js components.

When you write `require or define(['jquery'], ($) => {})`, breeze will pass jquery as $ in callback because 'jquery' is registered in breezemap: https://github.com/breezefront/module-breeze/blob/master/view/frontend/web/js/core/define.js#L15

So, to get this code work as expected in breeze:

![image](https://github.com/breezefront/community/assets/306080/2947abc2-4f7e-4a5e-8a6c-acd412ee94df)

you need to add `mageUtils` to breezemap.

**Reply by IbrahimS2** [reply]
@vovayatsyuk  It might be a good idea to have this added to Breeze as part of the core which will help with lots of other integrations, what do you think?

**Reply by IbrahimS2** [reply]
@vovayatsyuk A question, I refactored a JS so it uses the $.sections to access the customer data, anyhow, there is an issue where it returns undefined even though I am certain it is proper, please let me know if you have any clue:

```
define([
    'jquery',
    'uiComponent',
    'Magento_Customer/js/customer-data',
    'Magento_Customer/js/view/customer',
    'mage/validation'
], function ($, Component, customerData) {
    return Component.extend({
        initialize: function (config) {
            this.faq = $.sections.get('faq');
            this._super();
        },
        saveData: function () {
            $.sections.set('faq', this.faq());
        },
        initObservable: function () {
            console.log(this.faq); // undefined, no clue!!?
            
            this._super()
                .observe({
                    title: this.faq().title,
                    email: this.faq().email,
                    name: this.faq().name,
                    isEmailVisible: null,
                    isNameVisible: null,
                    isNotificationVisible: null,
                    isNotificationChecked: null
                });

            this.title.subscribe(function (value) {
                this.faq().title = value;
                this.saveData();
            }.bind(this));
            this.email.subscribe(function (value) {
                this.faq().email = value;
                this.saveData();
            }.bind(this));
            this.name.subscribe(function (value) {
                this.faq().name = value;
                this.saveData();
            }.bind(this));
            this.faq.subscribe(function (value) {
                if (this.name() != value.name) {
                    this.name(value.name)
                }
                if (this.email() != value.email) {
                    this.email(value.email)
                }
                if (this.title() != value.title) {
                    this.title(value.title)
                }
            }.bind(this));

            return this;
        },
        toggleEmail: function (element) {
            this.isEmailVisible(element.checked);
            this.isNotificationChecked(element.checked);
        },
        toggleNameAndNotification: function (element) {
            this.isNameVisible(element.checked);
            this.isNotificationVisible(element.checked);
            this.isEmailVisible(element.checked && this.isNotificationChecked());
        },
        formSubmit: function (form) {
            if (!$(form).validation() || !$(form).validation('isValid')) {
                return false;
            }
            $(form).find(':input').attr('readonly', true);
            $(form).find('button').attr('disabled', true);

            return true;
        }
    });
});

```

Please note component is already defined on another file when importing:

```
define(["uiComponent"], function (Component) {
    "use strict";

    Component.register("Amasty_Faq/js/askform");
});
```

**Reply by vovayatsyuk** [reply]
Try to set the value in initObservable function. In this case, you don't need to override `initialize` method:

```js
// initialize: function (config) {
//     this.faq = $.sections.get('faq');
//     this._super();
// },
initObservable: function () {
    this.faq = ...
```

p.s. The issue is caused by different order of function calls in Breeze:

- In Luma initObservable is called from `initialize` method (when you call `this._super();`). 
- In Breeze, initObservable is called before `initialize`.

I think I'm gonna fix this in the next Breeze update

p.p.s Still, I don't understand why Amasty created the `initialize` method to declare observable - why don't they move that code into `initObservable` just like I show?

**Reply by IbrahimS2** [reply]
@vovayatsyuk No clue, it just came to my attention this is the way it works with Luma; anyhow should I wait for this update if it will fix it on its own?

**Reply by IbrahimS2** [reply]
@vovayatsyuk Let me know, please.

**Comment by MaxSouza** [comment]
Hey @IbrahimS2,  Did you get it fully integrated with Breeze?

**Reply by IbrahimS2** [reply]
@MaxSouza 90% yes

**Reply by MaxSouza** [reply]
@IbrahimS2  Do you plan to make it available as a Breeze extension?

**Reply by IbrahimS2** [reply]
@MaxSouza 100%, I will be releasing it by next Monday
